cpp/oneapi/dal/algo/decision_forest/backend/gpu/feature_type_helper_dpc.cpp

```diff
@@ -0,0 +1,329 @@
+        h.depends_on(deps);
+        h.parallel_for(range, [=](sycl::id<1> idx) {
+            full_data[idx * column_count + column_idx] = column_data[idx];
+        });
+    });
+
+    return event;
+}
+
+template <typename Float, typename Bin, typename Index>
+indexed_features<Float, Bin, Index>::indexed_features(sycl::queue& q,
+                                                      std::int64_t min_bin_size,
+                                                      std::int64_t max_bins)
+        : queue_(q) {
+    min_bin_size_ = de::integral_cast<Index>(min_bin_size);
+    max_bins_ = de::integral_cast<Index>(max_bins);
+}
+
+template <typename Float, typename Bin, typename Index>
+std::int64_t indexed_features<Float, Bin, Index>::get_required_mem_size(std::int64_t row_count,
+                                                                        std::int64_t column_count,
+                                                                        std::int64_t max_bins) {
+    std::int64_t requiredMem = 0;
+    requiredMem += sizeof(Bin) * (column_count + 1); // bin_offsets
+    requiredMem +=
+        sizeof(Bin) * row_count * column_count; // data vs ftrs bin map table (full_data_nd)
+    requiredMem +=
+        sizeof(Float) * column_count * std::min(max_bins, row_count); // bin_borders for each column
+
+    return requiredMem;
+}
+
+template <typename Float, typename Bin, typename Index>
+sycl::event indexed_features<Float, Bin, Index>::operator()(
+    const table& tbl,
+    const dal::backend::event_vector& deps) {
+    sycl::event::wait_and_throw(deps);
+
+    const auto data_nd_ =
+        pr::flatten_table<Float, row_accessor>(queue_, tbl, sycl::usm::alloc::device);
+
+    if (tbl.get_row_count() > de::limits<Index>::max()) {
+        throw domain_error(dal::detail::error_messages::invalid_range_of_rows());
+    }
+    if (tbl.get_column_count() > de::limits<Index>::max()) {
+        throw domain_error(dal::detail::error_messages::invalid_range_of_columns());
+    }
+
+    row_count_ = de::integral_cast<Index>(tbl.get_row_count());
+    column_count_ = de::integral_cast<Index>(tbl.get_column_count());
+    total_bins_ = 0;
+
+    //allocating buffers
+    full_data_nd_ =
+        pr::ndarray<Bin, 2>::empty(queue_, { row_count_, column_count_ }, sycl::usm::alloc::device);
+    bin_offsets_nd_ =
+        pr::ndarray<Index, 1>::empty(queue_, { column_count_ + 1 }, sycl::usm::alloc::device);
+
+    entries_.resize(column_count_);
+
+    auto values_nd = pr::ndarray<Float, 1>::empty(queue_, { row_count_ }, sycl::usm::alloc::device);
+    auto indices_nd =
+        pr::ndarray<Index, 1>::empty(queue_, { row_count_ }, sycl::usm::alloc::device);
+    std::vector<pr::ndarray<Bin, 1>> column_bin_vec_;
+    column_bin_vec_.resize(column_count_);
+
+    for (Index i = 0; i < column_count_; i++) {
+        column_bin_vec_[i] =
+            pr::ndarray<Bin, 1>::empty(queue_, { row_count_ }, sycl::usm::alloc::device);
+    }
```

Comment:
Suggest to have separate function for that.

BTW, why vector is required? We allocate `column_count_ x row_count_` elements in total. Can we allocate large buffer at once?

Notes regarding vector usage:
```c++
std::vector<pr::ndarray<Bin, 1>> column_bin_vec_{ column_count_ };
```

```c++
for (auto& bin : column_bin_vec_) {
   bin = pr::ndarray<Bin, 1>::empty(queue_, { row_count_ }, sycl::usm::alloc::device);
}
```